### PR TITLE
KHP3-5065 - LDL to Zero

### PR DESCRIPTION
**Pull request by PatrickWaweru**
KHP3-5065 - LDL to Zero

### CODE REVIEW

File: api/pom.xml [hunk 1]
```diff
@@ -5,7 +5,7 @@
 	<parent>
 		<groupId>org.openmrs.module</groupId>
 		<artifactId>kenyaemrorderentry</artifactId>
-		<version>1.6.2</version>
```
Comment (ojwanganto) [inline api/pom.xml:8]: this should be a snapshot still

File: api/pom.xml [hunk 2]
```diff
@@ -157,8 +157,14 @@
 				<directory>src/main/resources</directory>
 				<filtering>true</filtering>
 			</resource>
+			<resource>
+				<directory>src/main/unfiltered-resources</directory>
+                <includes>
```
Comment (ojwanganto) [inline api/pom.xml:162]: what exactly are the new lines adding?

File: api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java [hunk 3]
```diff
@@ -638,8 +638,10 @@ private void updateOrder(Integer orderId, String result, String specimenStatus,
 
                     if (getSystemType() == ModuleConstants.CHAI_SYSTEM) {
                         if (result.equalsIgnoreCase(lDLResult) || result.equalsIgnoreCase(labwarelDLResult) || result.contains("LDL")) {
-                            conceptToRetain = vlTestConceptQualitative;
-                            o.setValueCoded(LDLConcept);
+                            // conceptToRetain = vlTestConceptQualitative;
+                            // o.setValueCoded(LDLConcept);
+                            conceptToRetain = vlTestConceptQuantitative;
```
Comment (ojwanganto) [inline api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java:643]: I think we can rename this variable appropriately. We'll only have one viral load concept going forward.

File: api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java [hunk 4]
```diff
@@ -651,8 +653,10 @@ private void updateOrder(Integer orderId, String result, String specimenStatus,
                     } else if (getSystemType() == ModuleConstants.LABWARE_SYSTEM) {
                         result = result.toLowerCase().trim(); // convert to lowercase and trim
                         if (result.equalsIgnoreCase(lDLResult) || result.equalsIgnoreCase(labwarelDLResult) || result.contains("LDL")) {
-                            conceptToRetain = vlTestConceptQualitative;
-                            o.setValueCoded(LDLConcept);
+                            // conceptToRetain = vlTestConceptQualitative;
+                            // o.setValueCoded(LDLConcept);
+                            conceptToRetain = vlTestConceptQuantitative;
```
Comment (ojwanganto) [inline api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java:658]: same as above

File: api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java [hunk 5]
```diff
@@ -676,8 +680,10 @@ private void updateOrder(Integer orderId, String result, String specimenStatus,
                     } else if (getSystemType() == ModuleConstants.EDARP_SYSTEM) {
                         // System.out.println("Got sample result as: " + result);
                         if (result.equalsIgnoreCase(lDLResult) || result.equalsIgnoreCase(labwarelDLResult) || result.contains("LDL")) {
-                            conceptToRetain = vlTestConceptQualitative;
-                            o.setValueCoded(LDLConcept);
+                            // conceptToRetain = vlTestConceptQualitative;
+                            // o.setValueCoded(LDLConcept);
+                            conceptToRetain = vlTestConceptQuantitative;
```
Comment (ojwanganto) [inline api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java:685]: same as above. I think just call it viral load

File: api/src/main/resources/liquibase.xml [hunk 6]
```diff
@@ -384,4 +384,17 @@
         </addColumn>
     </changeSet>
 
+    <!-- Add Order Entry stored procedures - VL to Zero -->
+    <changeSet author="pwaweru" id="kenyaemr_order_entry-sp_vl_LDL_to_Zero-20240215-153630" runOnChange="true">
```
Comment (ojwanganto) [inline api/src/main/resources/liquibase.xml:388]: I don't favor this approach. Wondering if we can resort to using a chore that executes the query. The chore will always just run once after which we can drop it in subsequent releases. Evaluate keenly

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/fragment/controller/manifest/ManifestFormFragmentController.java [hunk 7]
```diff
@@ -156,6 +162,42 @@ public EditManifestForm newEditManifestForm(@RequestParam(value = "manifestId",
         }
     }
 
+    public SimpleObject callLdlToZero() {
```
Comment (ojwanganto) [inline omod/src/main/java/org/openmrs/module/kenyaemrorderentry/fragment/controller/manifest/ManifestFormFragmentController.java:165]: You can extract this code into a chore and have it executed once during startup. You can drop the stored procedure once the process is done successfully

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/page/controller/orders/MaintenancePageController.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import org.openmrs.module.kenyaui.KenyaUiUtils;
+import org.openmrs.module.kenyaui.annotation.AppPage;
+import org.openmrs.scheduler.SchedulerService;
+import org.openmrs.ui.framework.SimpleObject;
+import org.openmrs.ui.framework.UiUtils;
+import org.openmrs.ui.framework.annotation.SpringBean;
+import org.openmrs.ui.framework.page.PageModel;
+
+@AppPage("kenyaemr.labmanifest")
+public class MaintenancePageController {
```
Comment (ojwanganto) [inline omod/src/main/java/org/openmrs/module/kenyaemrorderentry/page/controller/orders/MaintenancePageController.java:22]: We'll not need this.

File: omod/src/main/webapp/pages/orders/labOrdersCompleteResultManifestHome.gsp [hunk 9]
```diff
@@ -28,6 +28,7 @@
 
     def configuration = [
             [label: "Settings", iconProvider: "kenyaui", icon: "", label: "Settings", href: ui.pageLink("kenyaemrorderentry", "orders/settings")],
+            [label: "Maintenance", iconProvider: "kenyaui", icon: "", label: "Maintenance", href: ui.pageLink("kenyaemrorderentry", "orders/maintenance")],
```
Comment (ojwanganto) [inline omod/src/main/webapp/pages/orders/labOrdersCompleteResultManifestHome.gsp:31]: We can clean this

File: omod/src/main/webapp/pages/orders/labOrdersCompleteWithErrorResultsManifestHome.gsp [hunk 10]
```diff
@@ -28,6 +28,7 @@
 
     def configuration = [
             [label: "Settings", iconProvider: "kenyaui", icon: "", label: "Settings", href: ui.pageLink("kenyaemrorderentry", "orders/settings")],
+            [label: "Maintenance", iconProvider: "kenyaui", icon: "", label: "Maintenance", href: ui.pageLink("kenyaemrorderentry", "orders/maintenance")],
```
Comment (ojwanganto) [inline omod/src/main/webapp/pages/orders/labOrdersCompleteWithErrorResultsManifestHome.gsp:31]: ...and this

File: omod/src/main/webapp/pages/orders/maintenance.gsp [hunk 11]
```diff
@@ -0,0 +1,140 @@
+<%
```
Comment (ojwanganto) [inline omod/src/main/webapp/pages/orders/maintenance.gsp:1]: we'll not need this

**Review by ojwanganto — CHANGES_REQUESTED** [review]

**Comment by ojwanganto** [comment]
Please start handling the changes. I will give you a call for a walkthrough

File: api/pom.xml (hunk 1 above)
Comment (PatrickWaweru) [reply api/pom.xml:8]: Ok. Fixed.

File: api/pom.xml (hunk 2 above)
Comment (PatrickWaweru) [reply api/pom.xml:162]: New SQL script as a resource. It should not be filtered as filtering might modify the script.

File: api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java (hunk 3 above)
Comment (PatrickWaweru) [reply api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java:643]: I agree. It is work in progress.

File: api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java (hunk 4 above)
Comment (PatrickWaweru) [reply api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java:658]: noted

File: api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java (hunk 5 above)
Comment (PatrickWaweru) [reply api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java:685]: noted

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/page/controller/orders/MaintenancePageController.java (hunk 8 above)
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/page/controller/orders/MaintenancePageController.java:22]: Agreed

File: omod/src/main/webapp/pages/orders/labOrdersCompleteResultManifestHome.gsp (hunk 9 above)
Comment (PatrickWaweru) [reply omod/src/main/webapp/pages/orders/labOrdersCompleteResultManifestHome.gsp:31]: Ok

File: omod/src/main/webapp/pages/orders/maintenance.gsp (hunk 11 above)
Comment (PatrickWaweru) [reply omod/src/main/webapp/pages/orders/maintenance.gsp:1]: Agreed

File: omod/src/main/webapp/pages/orders/labOrdersCompleteWithErrorResultsManifestHome.gsp (hunk 10 above)
Comment (PatrickWaweru) [reply omod/src/main/webapp/pages/orders/labOrdersCompleteWithErrorResultsManifestHome.gsp:31]: Ok

File: api/src/main/resources/liquibase.xml (hunk 6 above)
Comment (PatrickWaweru) [reply api/src/main/resources/liquibase.xml:388]: Ok

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/fragment/controller/manifest/ManifestFormFragmentController.java (hunk 7 above)
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/fragment/controller/manifest/ManifestFormFragmentController.java]: Done: https://github.com/palladiumkenya/openmrs-module-kenyaemrorderentry/pull/297/commits/52271d0e14606409ecca71a24c0faa1d2006f5dc

File: omod/src/main/java/org/openmrs/module/kenyaemrorderentry/page/controller/orders/MaintenancePageController.java (hunk 8 above)
Comment (PatrickWaweru) [reply omod/src/main/java/org/openmrs/module/kenyaemrorderentry/page/controller/orders/MaintenancePageController.java]: Done: https://github.com/palladiumkenya/openmrs-module-kenyaemrorderentry/pull/297/commits/52271d0e14606409ecca71a24c0faa1d2006f5dc

File: omod/src/main/webapp/pages/orders/labOrdersCompleteResultManifestHome.gsp (hunk 9 above)
Comment (PatrickWaweru) [reply omod/src/main/webapp/pages/orders/labOrdersCompleteResultManifestHome.gsp]: Done: https://github.com/palladiumkenya/openmrs-module-kenyaemrorderentry/pull/297/commits/52271d0e14606409ecca71a24c0faa1d2006f5dc

File: omod/src/main/webapp/pages/orders/labOrdersCompleteWithErrorResultsManifestHome.gsp (hunk 10 above)
Comment (PatrickWaweru) [reply omod/src/main/webapp/pages/orders/labOrdersCompleteWithErrorResultsManifestHome.gsp]: Done: https://github.com/palladiumkenya/openmrs-module-kenyaemrorderentry/pull/297/commits/52271d0e14606409ecca71a24c0faa1d2006f5dc

File: omod/src/main/webapp/pages/orders/maintenance.gsp (hunk 11 above)
Comment (PatrickWaweru) [reply omod/src/main/webapp/pages/orders/maintenance.gsp]: Done: https://github.com/palladiumkenya/openmrs-module-kenyaemrorderentry/pull/297/commits/52271d0e14606409ecca71a24c0faa1d2006f5dc

File: api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java (hunk 3 above)
Comment (PatrickWaweru) [reply api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java]: Done: https://github.com/palladiumkenya/openmrs-module-kenyaemrorderentry/pull/297/commits/b70e89b9f228cc05cf039ef61f44e240e5bb29b3

File: api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java (hunk 4 above)
Comment (PatrickWaweru) [reply api/src/main/java/org/openmrs/module/kenyaemrorderentry/labDataExchange/LabOrderDataExchange.java]: Done: https://github.com/palladiumkenya/openmrs-module-kenyaemrorderentry/pull/297/commits/b70e89b9f228cc05cf039ef61f44e240e5bb29b3

**Comment by ojwanganto** [comment]
Have you tested and confirmed that this works well?

**Review by ojwanganto — APPROVED** [review]

**Comment by ojwanganto** [comment]
Please test it in a large db and get to know how long it takes

**Pull request merged**